File: packages/openmemory-py/src/openmemory/temporal_graph/timeline.py
import time
import json
from typing import List, Dict, Any, Optional

from ..core.db import db, q
from .query import query_facts_at_time, format_fact

# Port of backend/src/temporal_graph/timeline.ts


async def get_subject_timeline(
    subject: str, predicate: Optional[str] = None, userId: Optional[str] = None, user_id: Optional[str] = None
) -> List[Dict[str, Any]]:
    """
    Get a chronological list of changes (creating/invalidated) for a subject.

    Args:
        subject: The subject entity to track.
        predicate: Optional predicate filter.
        userId: Owner user ID.

    Returns:
        Sorted list of TimelineEntry dicts.
    """
    userId = userId or user_id
    conds = ["subject = ?"]
    params: List[Any] = [subject]

    if userId:
        conds.append("user_id = ?")
        params.append(userId)
    else:
        conds.append("user_id IS NULL")

    if predicate:
        conds.append("predicate = ?")
        params.append(predicate)

    t = q.tables
    sql = f"""
        SELECT subject, predicate, object, confidence, valid_from as validFrom, valid_to as validTo
        FROM {t['temporal_facts']}
        WHERE {' AND '.join(conds)}
        ORDER BY validFrom ASC
    """
    rows = await db.async_fetchall(sql, tuple(params))
    return _build_timeline(rows)


def _build_timeline(rows: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Helper to convert rows into event-based timeline entries."""
    timeline = []
    for row in rows:
        timeline.append(
            {
                "timestamp": row["validFrom"],
                "subject": row["subject"],
                "predicate": row["predicate"],
                "object": row["object"],
                "confidence": row["confidence"],
                "change_type": "created",
            }
        )
        if row["validTo"]:
            timeline.append(
                {
                    "timestamp": row["validTo"],
                    "subject": row["subject"],
                    "predicate": row["predicate"],
                    "object": row["object"],
                    "confidence": row["confidence"],
                    "change_type": "invalidated",
                }
            )
    timeline.sort(key=lambda x: x["timestamp"])
    return timeline


async def get_predicate_timeline(
    predicate: str,
    start: Optional[int] = None,
    end: Optional[int] = None,
    userId: Optional[str] = None,
    user_id: Optional[str] = None,
) -> List[Dict[str, Any]]:
    """Get a timeline of all facts using a specific predicate."""
    userId = userId or user_id
    conds = ["predicate = ?"]
    params: List[Any] = [predicate]

    if userId:
        conds.append("user_id = ?")
        params.append(userId)
    else:
        conds.append("user_id IS NULL")

    if start is not None:
        conds.append("validFrom >= ?")
        params.append(start)
    if end is not None:
        conds.append("validFrom <= ?")
        params.append(end)

    t = q.tables
    sql = f"""
        SELECT subject, predicate, object, confidence, valid_from as validFrom, valid_to as validTo
        FROM {t['temporal_facts']}
        WHERE {' AND '.join(conds)}
        ORDER BY validFrom ASC
    """
    rows = await db.async_fetchall(sql, tuple(params))
    return _build_timeline(rows)


async def get_changes_in_window(
    start: int, end: int, subject: Optional[str] = None, userId: Optional[str] = None, user_id: Optional[str] = None
) -> List[Dict[str, Any]]:
    """Get all temporal changes that occurred within a specific time window."""
    userId = userId or user_id
    conds = []
    params: List[Any] = [start, end, start, end]  # from_ts, to_ts, from_ts, to_ts

    if subject:
        conds.append("subject = ?")
        params.append(subject)
    if userId:
        conds.append("user_id = ?")
        params.append(userId)

    where_sub = f"AND {' AND '.join(conds)}" if conds else ""

    t = q.tables
    sql = f"""
        SELECT subject, predicate, object, confidence, valid_from as validFrom, valid_to as validTo
        FROM {t['temporal_facts']}
        WHERE ((valid_from >= ? AND valid_from <= ?) OR (valid_to >= ? AND valid_to <= ?))
        {where_sub}
    """
    rows = await db.async_fetchall(sql, tuple(params))
    timeline = []

    for row in rows:
        if row["validFrom"] >= start and row["validFrom"] <= end:
            timeline.append(
                {
                    "timestamp": row["validFrom"],
                    "subject": row["subject"],
                    "predicate": row["predicate"],
                    "object": row["object"],
                    "confidence": row["confidence"],
                    "change_type": "created",
                }
            )
        if row["validTo"] and row["validTo"] >= start and row["validTo"] <= end:
            timeline.append(
                {
                    "timestamp": row["validTo"],
                    "subject": row["subject"],
                    "predicate": row["predicate"],
                    "object": row["object"],
                    "confidence": row["confidence"],
                    "change_type": "invalidated",
                }
            )

    timeline.sort(key=lambda x: x["timestamp"])
    return timeline


async def compare_time_points(
    subject: str, t1: int, t2: int, userId: Optional[str] = None, user_id: Optional[str] = None
) -> Dict[str, List[Dict[str, Any]]]:
    """Compare the state of a subject at two different points in time."""
    userId = userId or user_id
    user_clause = "AND user_id = ?" if userId else "AND user_id IS NULL"
    user_param = (userId,) if userId else ()  # type: ignore[assignment]

    t = q.tables
    sql1 = f"""
        SELECT id, user_id, subject, predicate, object, valid_from as validFrom, valid_to as validTo, confidence, last_updated, metadata
        FROM {t['temporal_facts']}
        WHERE subject = ?
        AND valid_from <= ? AND (valid_to IS NULL OR valid_to >= ?)
        {user_clause}
    """
    f1 = await db.async_fetchall(sql1, (subject, t1, t1) + user_param)  # type: ignore[arg-type]
    f2 = await db.async_fetchall(sql1, (subject, t2, t2) + user_param)  # type: ignore[arg-type]

    m1 = {r["predicate"]: r for r in f1}
    m2 = {r["predicate"]: r for r in f2}

    added = []
    removed = []
    changed = []
    unchanged = []

    for pred, fact2 in m2.items():
        fact1 = m1.get(pred)
        if not fact1:
            added.append(format_fact(fact2))
        elif fact1["object"] != fact2["object"] or fact1["id"] != fact2["id"]:
            changed.append({
                "before": format_fact(fact1),
                "after": format_fact(fact2)
            })
        else:
            unchanged.append(format_fact(fact2))

    for pred, fact1 in m1.items():
        if pred not in m2:
            removed.append(format_fact(fact1))

    return {"added": added, "removed": removed, "changed": changed, "unchanged": unchanged}


async def get_change_frequency(
    subject: str, predicate: str, window_days: int = 30, userId: Optional[str] = None, user_id: Optional[str] = None
) -> Dict[str, Any]:
    """Calculate the frequency of changes for a specific subject-predicate pair."""
    userId = userId or user_id
    now = int(time.time()*1000)
    start = now - (window_days * 86400000)

    user_clause = "AND user_id = ?" if userId else ""
    user_param = (userId,) if userId else ()

    t = q.tables
    sql = f"""
        SELECT valid_from as validFrom, valid_to as validTo
        FROM {t['temporal_facts']}
        WHERE subject = ? AND predicate = ?
        AND valid_from >= ?
        {user_clause}
        ORDER BY validFrom ASC
    """
    rows = await db.async_fetchall(sql, (subject, predicate, start) + user_param)

    total_changes = len(rows)
    total_dur = 0
    valid_count = 0

    for r in rows:
        if r["validTo"]:
            total_dur += r["validTo"] - r["validFrom"]
            valid_count += 1

    avg_dur = total_dur / valid_count if valid_count > 0 else 0
    rate = total_changes / window_days

    return {
        "predicate": predicate,
        "total_changes": total_changes,
        "avg_duration_ms": avg_dur,
        "change_rate_per_day": rate
    }


async def get_volatile_facts(
    subject: Optional[str] = None, limit: int = 10, userId: Optional[str] = None, user_id: Optional[str] = None
) -> List[Dict[str, Any]]:
    """Identify facts that change frequently (volatility analysis)."""
    userId = userId or user_id
    conds = []
    params = []
    if subject:
        conds.append("subject = ?")
        params.append(subject)
    if userId:
        conds.append("user_id = ?")
        params.append(userId)

    where = f"WHERE {' AND '.join(conds)}" if conds else ""

    t = q.tables
    sql = f"""
        SELECT subject, predicate, COUNT(*) as change_count, AVG(confidence) as avg_confidence
        FROM {t['temporal_facts']}
        {where}
        GROUP BY subject, predicate
        HAVING change_count > 1
        ORDER BY change_count DESC, avg_confidence ASC
        LIMIT ?
    """
    rows = await db.async_fetchall(sql, tuple(params + [limit]))
    return rows

File: packages/openmemory-js/src/temporal_graph/timeline.ts
/**
 * @file timeline.ts
 * @description Temporal Knowledge Graph Timeline Analytics for OpenMemory.
 * @audited 2026-01-19
 */
import { q, transaction, allAsync, TABLES } from "../core/db";
import { env } from "../core/cfg";
import { logger } from "../utils/logger";
import { normalizeUserId } from "../utils";
import { rowToFact } from "./query";
import { TemporalFact, TemporalFactRow, TimelineEntry } from "./types";

/**
 * Helper to convert fact rows into a sorted chronological timeline of events.
 */
async function rowsToTimelineEntries(rows: TemporalFactRow[]): Promise<TimelineEntry[]> {
    const timeline: TimelineEntry[] = [];

    for (const row of rows) {
        // Hydrate row to TemporalFact to handle decryption and normalization
        const fact = await rowToFact(row);

        // Creation event
        timeline.push({
            timestamp: Number(fact.validFrom),
            subject: fact.subject,
            predicate: fact.predicate,
            object: fact.object,
            confidence: fact.confidence,
            changeType: "created",
        });

        // Invalidation event (if applicable)
        if (fact.validTo) {
            timeline.push({
                timestamp: Number(fact.validTo),
                subject: fact.subject,
                predicate: fact.predicate,
                object: fact.object,
                confidence: fact.confidence,
                changeType: "invalidated",
            });
        }
    }

    // Sorting is necessary because different predicates for the same subject can overlap
    return timeline.sort((a, b) => a.timestamp - b.timestamp);
}

/**
 * Retrieves the chronological timeline of facts associated with a specific subject.
 */
export const getSubjectTimeline = async (
    subject: string,
    predicate?: string,
    userId?: string | null,
): Promise<TimelineEntry[]> => {
    const uid = normalizeUserId(userId);
    let rows: TemporalFactRow[];

    if (predicate) {
        if (env.verbose) logger.debug(`[TEMPORAL] Fetching optimized timeline for ${subject} / ${predicate}`);
        // Use optimized query from repository
        rows = await q.getFactsBySubjectAndPredicate.all(subject, predicate, uid) as TemporalFactRow[];
    } else {
        // Fallback to existing broad query
        rows = await q.getFactsBySubject.all(subject, 0, true, 10000, uid) as TemporalFactRow[];
    }

    return await rowsToTimelineEntries(rows);
};

export const getPredicateTimeline = async (
    predicate: string,
    from?: Date,
    to?: Date,
    userId?: string | null,
): Promise<TimelineEntry[]> => {
    const uid = normalizeUserId(userId);
    const fromTs = from?.getTime();
    const toTs = to?.getTime();

    const rows = await q.getFactsByPredicate.all(predicate, fromTs, toTs, uid) as TemporalFactRow[];
    return await rowsToTimelineEntries(rows);
};

/**
 * Retrieves all factual changes (creations and invalidations) within a specific time window.
 */
export const getChangesInWindow = async (
    from: Date,
    to: Date,
    subject?: string,
    userId?: string | null,
): Promise<TimelineEntry[]> => {
    const fromTs = from.getTime();
    const toTs = to.getTime();
    const uid = normalizeUserId(userId);

    const rows = await q.getChangesInWindow.all(fromTs, toTs, subject, uid) as TemporalFactRow[];

    // Construct timeline but filter events strictly within window (Repo returns facts overlapping window boundary events)
    const timeline: TimelineEntry[] = [];

    for (const row of rows) {
        const rowFrom = Number(row.validFrom);
        const rowTo = row.validTo ? Number(row.validTo) : null;

        if (rowFrom >= fromTs && rowFrom <= toTs) {
            timeline.push({
                timestamp: rowFrom,
                subject: row.subject,
                predicate: row.predicate,
                object: row.object,
                confidence: row.confidence,
                changeType: "created",
            });
        }

        if (rowTo && rowTo >= fromTs && rowTo <= toTs) {
            timeline.push({
                timestamp: rowTo,
                subject: row.subject,
                predicate: row.predicate,
                object: row.object,
                confidence: row.confidence,
                changeType: "invalidated",
            });
        }
    }

    return timeline.sort((a, b) => a.timestamp - b.timestamp);
};

/**
 * Compares factual state of a subject between two points in time.
 * Returns added, removed, changed, and unchanged facts.
 */
export const compareTimePoints = async (
    subject: string,
    time1: Date,
    time2: Date,
    userId?: string | null,
): Promise<{
    added: TemporalFact[];
    removed: TemporalFact[];
    changed: Array<{ before: TemporalFact; after: TemporalFact }>;
    unchanged: TemporalFact[];
}> => {
    const t1Ts = time1.getTime();
    const t2Ts = time2.getTime();
    const uid = normalizeUserId(userId);

    // Use transaction for consistency if desired, though snapshot isolation might not be guaranteed across two SELECTs in SQLite without serializable.
    // However, since we are reading historical data, it's likely stable unless active writes are happening.

    return await transaction.run(async () => {
        // Fetch facts at T1 using Repo
        const rowsT1 = await q.queryFactsAtTime.all(t1Ts, subject, undefined, undefined, 0, uid) as TemporalFactRow[];
        // Fetch facts at T2 using Repo
        const rowsT2 = await q.queryFactsAtTime.all(t2Ts, subject, undefined, undefined, 0, uid) as TemporalFactRow[];

        const mapT1 = new Map<string, TemporalFactRow>();
        const mapT2 = new Map<string, TemporalFactRow>();

        for (const f of rowsT1) mapT1.set(f.predicate, f);
        for (const f of rowsT2) mapT2.set(f.predicate, f);

        const added: TemporalFact[] = [];
        const removed: TemporalFact[] = [];
        const changed: Array<{ before: TemporalFact; after: TemporalFact }> = [];
        const unchanged: TemporalFact[] = [];

        // Find added and changed
        for (const [pred, fact2] of mapT2) {
            const fact1 = mapT1.get(pred);
            if (!fact1) {
                added.push(await rowToFact(fact2));
            } else if (fact1.object !== fact2.object || fact1.id !== fact2.id) {
                changed.push({
                    before: await rowToFact(fact1),
                    after: await rowToFact(fact2),
                });
            } else {
                unchanged.push(await rowToFact(fact2));
            }
        }

        // Find removed
        for (const [pred, fact1] of mapT1) {
            if (!mapT2.has(pred)) {
                removed.push(await rowToFact(fact1));
            }
        }

        return { added, removed, changed, unchanged };
    });
};

export const getChangeFrequency = async (
    subject: string,
    predicate: string,
    windowDays: number = 30,
    userId?: string | null,
): Promise<{
    predicate: string;
    totalChanges: number;
    avgDurationMs: number;
    changeRatePerDay: number;
}> => {
    const now = Date.now();
    const windowStart = now - windowDays * 86400000;
    const uid = normalizeUserId(userId);

    // Use getFactsByPredicate with validFrom filter
    const rows = await q.getFactsByPredicate.all(predicate, windowStart, undefined, uid) as TemporalFactRow[];

    // Filter by subject in memory (since getFactsByPredicate doesn't filter subject)
    const validRows = rows.filter(r => r.subject === subject);

    const totalChanges = validRows.length;
    let totalDuration = 0;
    let validDurations = 0;

    for (const row of validRows) {
        if (row.validTo) {
            totalDuration += Number(row.validTo) - Number(row.validFrom);
            validDurations++;
        } else {
            // Integrity: Count active fact from valid_from to now (virtual duration)
            totalDuration += now - Number(row.validFrom);
            validDurations++;
        }
    }

    const avgDurationMs =
        validDurations > 0 ? totalDuration / validDurations : 0;
    const changeRatePerDay = totalChanges / windowDays;

    return {
        predicate,
        totalChanges,
        avgDurationMs,
        changeRatePerDay,
    };
};

export const getVolatileFacts = async (
    subject?: string,
    limit: number = 10,
    userId?: string | null,
): Promise<
    Array<{
        subject: string;
        predicate: string;
        changeCount: number;
        avgConfidence: number;
    }>
> => {
    const uid = normalizeUserId(userId);
    const rows = await q.getVolatileFacts.all(subject, limit, uid) as { subject: string, predicate: string, changeCount: number, avgConfidence: number }[];

    return rows.map((row) => ({
        subject: row.subject,
        predicate: row.predicate,
        changeCount: row.changeCount,
        avgConfidence: row.avgConfidence,
    }));
};
